src/librustc_resolve/lib.rs

```diff
@@ -2945,20 +2973,20 @@ impl<'a, 'crateloader: 'a> Resolver<'a, 'crateloader> {
         crate_lint: CrateLint
     ) -> PathResolution {
         let segments = &path.segments.iter()
-            .map(|seg| seg.ident)
+            .map(|seg| (seg.ident, Some(seg.id)))
             .collect::<Vec<_>>();
         self.smart_resolve_path_fragment(id, qself, segments, path.span, source, crate_lint)
     }
 
     fn smart_resolve_path_fragment(&mut self,
                                    id: NodeId,
                                    qself: Option<&QSelf>,
-                                   path: &[Ident],
+                                   path: &[(Ident, Option<NodeId>)],
```

Comment:
Resolve code is a lot uglier now due to these `&[(Ident, Option<NodeId>)]` things, but hopefully this can be refactored later.

Reply:
Maybe a resolve-specific `PathSegment` struct would be better?